src/types.jl

```diff
@@ -18,42 +27,23 @@ example, the dimensions of velocity are `Dimensions(length=1, time=-1)`.
 - `luminosity`: luminosity dimension (i.e., cd^(luminosity))
 - `amount`: amount dimension (i.e., mol^(amount))
 """
-struct Dimensions{R <: Real}
+struct Dimensions{R<:Real} <: AbstractDimensions{R}
     length::R
     mass::R
     time::R
     current::R
     temperature::R
     luminosity::R
     amount::R
-
-    function Dimensions(length::_R,
-                        mass::_R,
-                        time::_R,
-                        current::_R,
-                        temperature::_R,
-                        luminosity::_R,
-                        amount::_R) where {_R<:Real}
-        new{_R}(length, mass, time, current, temperature, luminosity, amount)
-    end
-    Dimensions(; kws...) = Dimensions(DEFAULT_DIM_TYPE; kws...)
-    Dimensions(::Type{_R}; kws...) where {_R} = Dimensions(
-        tryrationalize(_R, get(kws, :length,      zero(_R))),
-        tryrationalize(_R, get(kws, :mass,        zero(_R))),
-        tryrationalize(_R, get(kws, :time,        zero(_R))),
-        tryrationalize(_R, get(kws, :current,     zero(_R))),
-        tryrationalize(_R, get(kws, :temperature, zero(_R))),
-        tryrationalize(_R, get(kws, :luminosity,  zero(_R))),
-        tryrationalize(_R, get(kws, :amount,      zero(_R))),
-    )
-    Dimensions{_R}(; kws...) where {_R} = Dimensions(_R; kws...)
-    Dimensions{_R}(args...) where {_R} = Dimensions(Base.Fix1(convert, _R).(args)...)
-    Dimensions{_R}(d::Dimensions) where {_R} = Dimensions{_R}(d.length, d.mass, d.time, d.current, d.temperature, d.luminosity, d.amount)
 end
 
-const DIMENSION_NAMES = Base.fieldnames(Dimensions)
-const DIMENSION_SYNONYMS = (:m, :kg, :s, :A, :K, :cd, :mol)
-const SYNONYM_MAPPING = NamedTuple(DIMENSION_NAMES .=> DIMENSION_SYNONYMS)
+(::Type{D})(::Type{R}; kws...) where {R,D<:AbstractDimensions} = D{R}((tryrationalize(R, get(kws, k, zero(R))) for k in fieldnames(D))...)
+(::Type{D})(; kws...) where {D<:AbstractDimensions} = D(DEFAULT_DIM_TYPE; kws...)
+
+(::Type{D})(args...) where {R,D<:AbstractDimensions{R}} = constructor_of(D)(Base.Fix1(convert, R).(args)...)
+(::Type{D})(; kws...) where {R,D<:AbstractDimensions{R}} = constructor_of(D)(R; kws...)
+(::Type{D})(d::AbstractDimensions) where {R,D<:AbstractDimensions{R}} = D((getfield(d, k) for k in fieldnames(D))...)
```

Comment:
Is building generic constructors like this a bad practice? It means the user doesn't have to write the constructors themselves (but could always overload them if needed).

Reply:
Seems like these constructors are much slower. I guess the compiler isn't able to inline some of the stuff here. (Maybe `fieldnames` is not inferrable. i.e., a `@generated function` is needed?)